File: src/ait/cli/commands/worktree.py
from __future__ import annotations

from ..shared import export_app_namespace

export_app_namespace(globals())


@worktree_app.command(
    "show",
    help="Inspect one registered worktree and its local workspace state.",
    short_help="Inspect one worktree.",
)
def worktree_show(name: Optional[str] = typer.Argument(None), json_output: bool = typer.Option(False, "--json")):
    try:
        data = local_get_worktree(_ctx(), name)
    except (KeyError, ValueError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@worktree_app.command(
    "open",
    help="Print a worktree path or shell-open helpers for entering it.",
    short_help="Print shell-open helpers.",
)
@worktree_app.command(
    "path",
    help="Print a worktree path or shell-open helpers for entering it.",
    short_help="Print a worktree path.",
)
def worktree_path(
    name: Optional[str] = typer.Argument(None),
    shell_output: bool = typer.Option(False, "--shell", help="Print shell commands that enter the worktree and export its runtime environment"),
    json_output: bool = typer.Option(False, "--json"),
):
    try:
        ctx = _ctx()
        data = local_get_worktree(ctx, name)
        touched = _touch_worktree_usage_safely(ctx, name=name)
        if touched is not None:
            data = touched
    except (KeyError, ValueError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    canonical_path = str(data.get("path") or "")
    open_path_value = str(data.get("open_path") or data.get("alias_path") or canonical_path)
    if not canonical_path:
        raise typer.BadParameter("Worktree has no registered path.")
    cd_command = f"cd {shlex.quote(open_path_value)}"
    runtime_paths = _worktree_runtime_paths(open_path_value)
    shell_command = _worktree_shell_command(open_path_value, data)
    if json_output:
        _emit(
            {
                "name": data.get("name"),
                "path": canonical_path,
                "open_path": open_path_value,
                "alias_path": data.get("alias_path"),
                "cd_command": cd_command,
                "shell_command": shell_command,
                "current_line": data.get("current_line"),
                "workspace_status": data.get("workspace_status"),
                "changed_count": data.get("changed_count"),
                "src_path": runtime_paths.get("src_path"),
                "venv_path": runtime_paths.get("venv_path"),
                "venv_bin_path": runtime_paths.get("venv_bin_path"),
            },
            True,
        )
        return
    typer.echo(shell_command if shell_output else open_path_value)


@worktree_app.command(
    "exec",
    help="Run one command inside a named worktree from the repo root.",
    short_help="Run a command in a worktree.",
)
def worktree_exec(
    name: str = typer.Argument(..., help="Registered worktree name"),
    command: list[str] = typer.Argument(..., help="Command to run inside the worktree. Pass it after -- when forwarding flags."),
    json_output: bool = typer.Option(False, "--json"),
):
    try:
        ctx = _ctx()
        data = local_get_worktree(ctx, name)
    except (KeyError, ValueError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _touch_worktree_usage_safely(ctx, name=name)
    path_value = str(data.get("path") or "")
    if not path_value:
        raise typer.BadParameter("Worktree has no registered path.")
    worktree_path = Path(path_value)
    if not worktree_path.is_dir():
        raise typer.BadParameter(f"Worktree path is missing: {path_value}")
    env = _worktree_runtime_env(path_value, {"name": data.get("name") or name, "current_line": data.get("current_line")})
    try:
        completed = subprocess.run(command, cwd=path_value, env=env, capture_output=True, text=True)
    except FileNotFoundError as exc:
        raise typer.BadParameter(f"Command not found: {command[0]}") from exc
    payload = {
        "name": data.get("name") or name,
        "path": path_value,
        "current_line": data.get("current_line"),
        "command": command,
        "returncode": completed.returncode,
        "stdout": completed.stdout,
        "stderr": completed.stderr,
    }
    if json_output:
        _emit(payload, True)
    else:
        if completed.stdout:
            typer.echo(completed.stdout, nl=False)
        if completed.stderr:
            typer.echo(completed.stderr, nl=False, err=True)
    if completed.returncode != 0:
        raise typer.Exit(code=completed.returncode)


@worktree_app.command(
    "doctor",
    help="Inspect registered worktree health and stale entries.",
    short_help="Check worktree health.",
)
def worktree_doctor_cmd(json_output: bool = typer.Option(False, "--json")):
    data = local_worktree_doctor(_ctx())
    if json_output:
        _emit(data, True)
        return
    _render_worktree_doctor(data)


@worktree_app.command(
    "cleanup-candidates",
    help="Inspect lifecycle-aware cleanup candidates without deleting anything.",
    short_help="List cleanup candidates.",
)
def worktree_cleanup_candidates(
    older_than: str = typer.Option("7d", "--older-than", help="Idle threshold like 7d, 12h, or 30m."),
    cleanup_policy: Optional[str] = typer.Option(None, "--policy", help="Filter to one cleanup policy."),
    allow_manual_only: bool = typer.Option(
        False,
        "--allow-manual-only",
        help="Also surface clean manual_only worktrees that are only blocked by explicit operator intent.",
    ),
    include_protected: bool = typer.Option(False, "--include-protected", help="Also list protected rows and their reasons."),
    json_output: bool = typer.Option(False, "--json"),
):
    try:
        data = local_list_worktree_cleanup_candidates(
            _ctx(),
            older_than=older_than,
            cleanup_policy=cleanup_policy,
            include_protected=include_protected,
            allow_manual_only=allow_manual_only,
        )
    except (KeyError, ValueError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    if json_output:
        _emit(data, True)
        return
    _render_worktree_cleanup_candidates(data)


@worktree_app.command(
    "cleanup",
    help="Remove lifecycle-approved cleanup candidates after an explicit confirmation.",
    short_help="Clean up worktrees.",
)
def worktree_cleanup(
    older_than: str = typer.Option("7d", "--older-than", help="Idle threshold like 7d, 12h, or 30m."),
    cleanup_policy: Optional[str] = typer.Option(None, "--policy", help="Filter to one cleanup policy."),
    allow_manual_only: bool = typer.Option(
        False,
        "--allow-manual-only",
        help="Explicitly include clean manual_only worktrees that are otherwise safe to remove.",
    ),
    limit: Optional[int] = typer.Option(None, "--limit", min=1, help="Only clean the first N selected candidates."),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview cleanup without deleting anything."),
    yes: bool = typer.Option(False, "--yes", help="Confirm lifecycle cleanup."),
    json_output: bool = typer.Option(False, "--json"),
):
    if not dry_run and not yes:
        raise typer.BadParameter("Pass --yes to apply worktree cleanup, or use --dry-run to preview it.")
    ctx = _ctx()
    try:
        data = _run_locked_workspace_command(
            ctx,
            "worktree cleanup",
            lambda: local_cleanup_worktrees(
                ctx,
                older_than=older_than,
                cleanup_policy=cleanup_policy,
                allow_manual_only=allow_manual_only,
                limit=limit,
                dry_run=dry_run,
            ),
        )
    except (KeyError, ValueError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    if json_output:
        _emit(data, True)
        return
    _render_worktree_cleanup_report(data)


@worktree_app.command(
    "prune-stale",
    help="Remove stale worktree registrations that no longer exist on disk.",
    short_help="Prune stale worktrees.",
)
def worktree_prune_stale(
    dry_run: bool = typer.Option(False, "--dry-run", help="Show which stale worktree registrations would be removed"),
    json_output: bool = typer.Option(False, "--json"),
):
    data = local_prune_stale_worktrees(_ctx(), dry_run=dry_run)
    if json_output:
        _emit(data, True)
        return
    _render_worktree_prune_report(data)


@worktree_app.command(
    "list",
    help="List registered worktrees. Use --refresh to verify current filesystem status before printing.",
    short_help="List worktrees.",
)
def worktree_list(
    json_output: bool = typer.Option(False, "--json"),
    refresh: bool = typer.Option(
        False,
        "--refresh",
        help="Verify each worktree against the live filesystem before printing and refresh the cached status fields.",
    ),
):
    rows = local_list_worktrees(_ctx(), refresh_status=refresh)
    if json_output:
        _emit(rows, True)
        return
    _render_worktrees(rows)


@worktree_app.command(
    "sync",
    help="Restore one worktree, or all live worktrees, to their intended line heads.",
    short_help="Sync worktrees.",
)
def worktree_sync(
    name: Optional[str] = typer.Argument(None),
    all_worktrees: bool = typer.Option(False, "--all", help="Sync every live registered worktree to its own current line"),
    line_name: Optional[str] = typer.Option(None, "--line", help="Restore a specific line head into the target worktree"),
    force: bool = typer.Option(False, "--force", help="Overwrite unsaved changes inside the target worktree"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview restore actions without writing workspace files"),
    json_output: bool = typer.Option(False, "--json"),
):
    if all_worktrees and name is not None:
        raise typer.BadParameter("Choose either a worktree name or --all")
    if all_worktrees and line_name is not None:
        raise typer.BadParameter("--line cannot be combined with --all; each worktree syncs to its own current line")
    ctx = _ctx()
    try:
        data = _run_locked_workspace_command(
            ctx,
            "worktree sync",
            lambda: local_sync_all_worktrees(ctx, force=force, dry_run=dry_run)
            if all_worktrees
            else local_sync_worktree(ctx, name, line_name=line_name, force=force, dry_run=dry_run),
        )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    if json_output:
        _emit(data, True)
    elif all_worktrees:
        _render_worktree_sync_report(data)
    else:
        _emit(data, False)
    if all_worktrees and not data.get("ok", True):
        raise typer.Exit(code=2)


@worktree_app.command(
    "recreate",
    help="Recreate one missing task-bound worktree from durable local/remote lineage.",
    short_help="Recreate a missing worktree.",
)
def worktree_recreate(
    name: Optional[str] = typer.Argument(None),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview the recreate source and target paths without writing files"),
    json_output: bool = typer.Option(False, "--json"),
):
    ctx = _ctx()
    try:
        data = _run_locked_workspace_command(
            ctx,
            "worktree recreate",
            lambda: local_recreate_worktree(ctx, name, dry_run=dry_run),
        )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@worktree_app.command(
    "restore-owned-head",
    help="Restore one task-bound worktree to the last clean owned snapshot before foreign lineage contamination.",
    short_help="Restore a clean owned head.",
)
def worktree_restore_owned_head(
    name: Optional[str] = typer.Argument(None),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview the restore-owned-head plan without changing the workspace"),
    json_output: bool = typer.Option(False, "--json"),
):
    ctx = _ctx()
    try:
        data = _run_locked_workspace_command(
            ctx,
            "worktree restore-owned-head",
            lambda: local_restore_owned_head(ctx, name, dry_run=dry_run),
        )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@worktree_app.command(
    "rebase",
    help="Retarget one task worktree onto a newer base line without reopening the worktree path.",
    short_help="Rebase a worktree.",
)
def worktree_rebase(
    name: Optional[str] = typer.Argument(None),
    onto_line: Optional[str] = typer.Option(None, "--onto", help="Base line whose current head should become the new fork point"),
    continue_rebase: bool = typer.Option(False, "--continue", help="Finalize a conflicted worktree rebase after resolving files"),
    abort_rebase: bool = typer.Option(False, "--abort", help="Abort a conflicted worktree rebase and restore the original head"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview the rebase plan without changing the workspace"),
    json_output: bool = typer.Option(False, "--json"),
):
    if continue_rebase and abort_rebase:
        raise typer.BadParameter("Choose either --continue or --abort")
    if dry_run and (continue_rebase or abort_rebase):
        raise typer.BadParameter("--dry-run cannot be combined with --continue or --abort")
    ctx = _ctx()
    try:
        if continue_rebase:
            data = _run_locked_workspace_command(
                ctx,
                "worktree rebase --continue",
                lambda: local_continue_worktree_rebase(ctx, name),
            )
        elif abort_rebase:
            data = _run_locked_workspace_command(
                ctx,
                "worktree rebase --abort",
                lambda: local_abort_worktree_rebase(ctx, name),
            )
        else:
            data = _run_locked_workspace_command(
                ctx,
                "worktree rebase",
                lambda: local_preview_worktree_rebase(ctx, name, onto_line_name=onto_line)
                if dry_run
                else local_rebase_worktree(ctx, name, onto_line_name=onto_line),
            )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@worktree_app.command(
    "remove",
    help="Remove one or more worktree registrations and optionally delete their paths.",
    short_help="Remove worktrees.",
)
def worktree_remove(
    names: list[str] = typer.Argument(None),
    all_stale: bool = typer.Option(False, "--all-stale", help="Remove every stale worktree registration from the registry"),
    delete_path: bool = typer.Option(False, "--delete-path", help="Also delete the worktree directory after unregistering it"),
    force: bool = typer.Option(False, "--force", help="Allow removing a dirty worktree"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview the selected worktree removals without writing changes"),
    json_output: bool = typer.Option(False, "--json"),
):
    if all_stale and names:
        raise typer.BadParameter("Choose either one or more worktree names or --all-stale")
    if not all_stale and not names:
        raise typer.BadParameter("Provide one or more worktree names or use --all-stale")
    if all_stale and delete_path:
        raise typer.BadParameter("--delete-path cannot be combined with --all-stale")
    if all_stale and force:
        raise typer.BadParameter("--force cannot be combined with --all-stale")
    ctx = _ctx()
    try:
        data = _run_locked_workspace_command(
            ctx,
            "worktree remove",
            lambda: local_prune_stale_worktrees(ctx, dry_run=dry_run)
            if all_stale
            else local_remove_worktrees(ctx, names, delete_path=delete_path, force=force, dry_run=dry_run)
            if len(names) > 1 or dry_run
            else local_remove_worktree(ctx, names[0], delete_path=delete_path, force=force),
        )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    if json_output:
        _emit(data, True)
        return
    if all_stale:
        _render_worktree_prune_report(data)
        return
    _emit(data, False)

File: ait-core/rust/crates/ait-cli/src/main_app/worktree.rs
const WORKTREE_DESTRUCTIVE_CONFIRMATION_ERROR: &str =
    "Pass --yes to apply this destructive worktree operation, or use --dry-run to preview it.";

fn require_worktree_destructive_confirmation(dry_run: bool, yes: bool) -> Result<(), String> {
    if !dry_run && !yes {
        return Err(WORKTREE_DESTRUCTIVE_CONFIRMATION_ERROR.to_string());
    }
    Ok(())
}

fn run_worktree(repo: RepoRuntime, command: WorktreeCommand) -> Result<ExitCode, String> {
    match command {
        WorktreeCommand::Status(args) => {
            let payload = worktree_status(
                &repo,
                args.name.as_deref(),
                args.snapshot_id.as_deref(),
                args.line_name.as_deref(),
            )?;
            emit_worktree_status_result(&payload, args.json, args.verbose)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Restore(args) => {
            let payload = run_locked_workspace_command(&repo, "ait-cli worktree restore", || {
                worktree_restore(
                    &repo,
                    args.name.as_deref(),
                    args.snapshot_id.as_deref(),
                    args.line_name.as_deref(),
                    &args.paths,
                    args.force,
                    args.dry_run,
                )
            })?;
            emit_result(
                "ait-cli worktree restore",
                &payload,
                args.json,
                &[
                    "worktree_name",
                    "workspace_root",
                    "current_line",
                    "line_name",
                    "line_head_snapshot_id",
                    "target_snapshot_id",
                    "applied",
                ],
            )?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Show(args) => {
            let payload = worktree_get(&repo, args.name.as_deref(), true)?;
            emit_worktree_show_result(&payload, args.json)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Path(args) => {
            let payload = touch_worktree_payload(&repo, args.name.as_deref())?;
            if args.json {
                print_json(&worktree_path_payload(&payload)?)?;
            } else if args.shell_output {
                println!(
                    "{}",
                    required_object_string_field(&payload, "shell_command")?
                );
            } else {
                println!("{}", worktree_open_path_text(&payload)?);
            }
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Doctor(args) => {
            let payload = worktree_doctor(&repo, args.refresh)?;
            emit_worktree_doctor_result(&payload, args.json)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::CleanupCandidates(args) => {
            let payload = worktree_cleanup_candidates(
                &repo,
                Some(args.older_than.as_str()),
                args.cleanup_policy.as_deref(),
                args.include_protected,
                args.allow_manual_only,
            )?;
            emit_worktree_cleanup_candidates_result(&payload, args.json)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Cleanup(args) => {
            require_worktree_destructive_confirmation(args.dry_run, args.yes)?;
            let payload = run_locked_workspace_command(&repo, "ait-cli worktree cleanup", || {
                worktree_cleanup(
                    &repo,
                    Some(args.older_than.as_str()),
                    args.cleanup_policy.as_deref(),
                    args.allow_manual_only,
                    args.limit,
                    args.dry_run,
                )
            })?;
            emit_worktree_cleanup_report_result(&payload, args.json)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::PruneStale(args) => {
            require_worktree_destructive_confirmation(args.dry_run, args.yes)?;
            let payload = worktree_prune_stale(&repo, args.dry_run)?;
            emit_worktree_prune_result(&payload, args.json)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::List(args) => {
            let payload = worktree_list(&repo, args.refresh)?;
            emit_worktree_list_result(&payload, args.json)?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Sync(args) => {
            if args.all_worktrees && args.name.is_some() {
                return Err("Choose either a worktree name or --all".to_string());
            }
            if args.all_worktrees && args.line_name.is_some() {
                return Err(
                    "--line cannot be combined with --all; each worktree syncs to its own current line"
                        .to_string(),
                );
            }
            let payload = run_locked_workspace_command(&repo, "ait-cli worktree sync", || {
                if args.all_worktrees {
                    worktree_sync_all(&repo, args.force, args.dry_run)
                } else {
                    worktree_sync(
                        &repo,
                        args.name.as_deref(),
                        args.line_name.as_deref(),
                        args.force,
                        args.dry_run,
                    )
                }
            })?;
            emit_worktree_sync_result(&payload, args.json)?;
            let exit = if args.all_worktrees
                && !payload
                    .get("ok")
                    .and_then(JsonValue::as_bool)
                    .unwrap_or(true)
            {
                ExitCode::from(2)
            } else {
                ExitCode::SUCCESS
            };
            Ok(exit)
        }
        WorktreeCommand::Recreate(args) => {
            let payload = run_locked_workspace_command(&repo, "ait-cli worktree recreate", || {
                worktree_recreate(&repo, args.name.as_deref(), args.dry_run)
            })?;
            emit_result(
                "ait-cli worktree recreate",
                &payload,
                args.json,
                &[
                    "name",
                    "path",
                    "status",
                    "current_line",
                    "target_base_line",
                    "fork_snapshot_id",
                ],
            )?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::RecoverTask(args) => {
            let payload =
                run_locked_workspace_command(&repo, "ait-cli worktree recover-task", || {
                    worktree_recover_task(
                        &repo,
                        &args.task_id,
                        &args.change,
                        args.remote.as_deref(),
                        args.dry_run,
                    )
                })?;
            emit_result(
                "ait-cli worktree recover-task",
                &payload,
                args.json,
                &[
                    "status",
                    "task_id",
                    "change_id",
                    "name",
                    "path",
                    "current_line",
                ],
            )?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::RestoreOwnedHead(args) => {
            let payload =
                run_locked_workspace_command(&repo, "ait-cli worktree restore-owned-head", || {
                    worktree_restore_owned_head(&repo, args.name.as_deref(), args.dry_run)
                })?;
            emit_result(
                "ait-cli worktree restore-owned-head",
                &payload,
                args.json,
                &[
                    "name",
                    "path",
                    "status",
                    "restored_snapshot_id",
                    "owned_head_snapshot_id",
                    "current_line",
                ],
            )?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Rebase(args) => {
            if args.continue_rebase && args.abort_rebase {
                return Err("Choose either --continue or --abort".to_string());
            }
            if args.dry_run && (args.continue_rebase || args.abort_rebase) {
                return Err("--dry-run cannot be combined with --continue or --abort".to_string());
            }
            let payload = if args.continue_rebase {
                run_locked_workspace_command(&repo, "ait-cli worktree rebase --continue", || {
                    worktree_continue_rebase(&repo, args.name.as_deref())
                })?
            } else if args.abort_rebase {
                run_locked_workspace_command(&repo, "ait-cli worktree rebase --abort", || {
                    worktree_abort_rebase(&repo, args.name.as_deref())
                })?
            } else {
                run_locked_workspace_command(&repo, "ait-cli worktree rebase", || {
                    if args.dry_run {
                        worktree_preview_rebase(
                            &repo,
                            args.name.as_deref(),
                            args.onto_line.as_deref(),
                        )
                    } else {
                        worktree_rebase(&repo, args.name.as_deref(), args.onto_line.as_deref())
                    }
                })?
            };
            emit_result(
                "ait-cli worktree rebase",
                &payload,
                args.json,
                &[
                    "name",
                    "path",
                    "status",
                    "current_line",
                    "target_base_line",
                    "rebase_state",
                    "conflict_count",
                ],
            )?;
            Ok(ExitCode::SUCCESS)
        }
        WorktreeCommand::Remove(args) => {
            if args.all_stale && !args.names.is_empty() {
                return Err("Choose either one or more worktree names or --all-stale".to_string());
            }
            if !args.all_stale && args.names.is_empty() {
                return Err("Provide one or more worktree names or use --all-stale".to_string());
            }
            if args.all_stale && args.delete_path {
                return Err("--delete-path cannot be combined with --all-stale".to_string());
            }
            if args.all_stale && args.force {
                return Err("--force cannot be combined with --all-stale".to_string());
            }
            require_worktree_destructive_confirmation(args.dry_run, args.yes)?;
            let payload = run_locked_workspace_command(&repo, "ait-cli worktree remove", || {
                if args.all_stale {
                    worktree_prune_stale(&repo, args.dry_run)
                } else {
                    worktree_remove(
                        &repo,
                        &args.names,
                        false,
                        args.delete_path,
                        args.force,
                        args.dry_run,
                    )
                }
            })?;
            if args.all_stale {
                emit_worktree_prune_result(&payload, args.json)?;
            } else if args.json {
                print_json(&payload)?;
            } else if args.dry_run || args.names.len() > 1 {
                emit_worktree_cleanup_rows(
                    "ait-cli worktree remove",
                    &payload,
                    "planned_count",
                    "planned_rows",
                    "removed_count",
                    "removed_rows",
                )?;
            } else {
                let row = payload
                    .get("removed_rows")
                    .and_then(JsonValue::as_array)
                    .and_then(|rows| rows.first())
                    .ok_or("worktree remove payload is missing removed_rows[0].")?;
                emit_result(
                    "ait-cli worktree remove",
                    row,
                    false,
                    &[
                        "name",
                        "path",
                        "deleted_path",
                        "alias_path",
                        "target_base_line",
                    ],
                )?;
            }
            Ok(ExitCode::SUCCESS)
        }
    }
}
